test/test_datapipe.py

```diff
@@ -2755,5 +2759,161 @@ def reset(self):
+        datapipe._snapshot_state = _SnapshotState.Deserialized
+        if rng is None:
+            rng = torch.Generator()
+        initial_rng_state = rng.get_state()
+        torch.utils.data.graph_settings.apply_shuffle_seed(datapipe, rng)
+        expected_full_res = list(datapipe)
+
+        # The next 3 lines simulate the effect of starting another epoch
+        # Then the reminder test cases show that you can restore to first epoch if needed
+        # as long as you have the initial RNG state
+        rng = rng.manual_seed(100)
+        torch.utils.data.graph_settings.apply_shuffle_seed(datapipe, rng)
+        list(datapipe)
+
+        # Test Case: fast forward works with list
+        datapipe.reset()
+        datapipe._snapshot_state = _SnapshotState.Deserialized
+        rng.set_state(initial_rng_state)
+        fast_forward_fn(datapipe, n_iterations, rng)
+        actual_res = list(datapipe)
+        self.assertEqual(len(datapipe) - n_iterations, len(actual_res))
+        self.assertEqual(expected_full_res[n_iterations:], actual_res)
+
+        # Test Case: fast forward works with iterator
+        datapipe.reset()
+        datapipe._snapshot_state = _SnapshotState.Deserialized
+        rng.set_state(initial_rng_state)
+        fast_forward_fn(datapipe, n_iterations, rng)
+        it = iter(datapipe)
+        actual_res = list(it)
+        self.assertEqual(len(datapipe) - n_iterations, len(actual_res))
+        self.assertEqual(expected_full_res[n_iterations:], actual_res)
+        with self.assertRaises(StopIteration):
+            next(it)
+
+    def test_simple_snapshot_graph(self):
+        graph1 = dp.iter.IterableWrapper(range(10))
+        self._fast_forward_graph_test_helper(graph1, _simple_graph_snapshot_restoration)
+
+        graph2 = graph1.map(_mul_10)
+        self._fast_forward_graph_test_helper(graph2, _simple_graph_snapshot_restoration)
+
+        rng = torch.Generator()
+        graph3 = graph2.shuffle()
+        self._fast_forward_graph_test_helper(graph3, _simple_graph_snapshot_restoration, rng=rng)
+
+        graph4 = graph3.map(_mul_10)
+        self._fast_forward_graph_test_helper(graph4, _simple_graph_snapshot_restoration, rng=rng)
+
+        graph5 = graph4.batch(2)
+        self._fast_forward_graph_test_helper(graph5, _simple_graph_snapshot_restoration, rng=rng)
+
+        # With `fork` and `zip`
+        cdp1, cdp2 = graph5.fork(2)
+        graph6 = cdp1.zip(cdp2)
+        self._fast_forward_graph_test_helper(graph6, _simple_graph_snapshot_restoration, rng=rng)
+
+        # With `fork` and `concat`
+        graph7 = cdp1.concat(cdp2)
+        self._fast_forward_graph_test_helper(graph7, _simple_graph_snapshot_restoration, rng=rng)
+
+        # Raises an exception if the graph has already been restored
+        with self.assertRaisesRegex(RuntimeError, "Snapshot restoration cannot be applied."):
+            _simple_graph_snapshot_restoration(graph7, 1)
```

Comment:
IIRC, what we have decided is we don't allow restoring from snapshot during iteration. It seems we are not even allowing to restore from snapshot multiple times even if the iteration has not started. Is this intentional?

Reply:
That's what I have in mind, because it doesn't to me why you would want to restore the same DataPipe (or graph) multiple times in a row? It seems like a potential user error?
But this is up for discussion and I'm open to change this and allow it to happen

Reply:
I am basically trying to align the behavior between DataPipe and DataLoader2: https://github.com/pytorch/data/blob/f629899aece23a15c155c7c7805b8db7a2bed2a7/torchdata/dataloader2/dataloader2.py#L176-L179

Reply:
So basically in DL2 you can load as many times as you want as long as you never created an iterator?
I guess my question is what is the use case for loading multiple times and not using it?

Reply:
Perhaps the middle ground is to allow it, but raises a warning instead? Because I think if someone is restoring multiple snapshots without iterating, they may be making a mistake.

Reply:
Since we are going to change DL2 to prevent this behavior, I am not going to make a change here (leaving the exception as it is). Let me know if I should still put a warning instead.